src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,416 @@
+                var indentation = indentationService.GetIndentation(document.WithText(formattedText), desiredCaretLine.LineNumber, cancellationToken);
+
+                // Insert whitespace for the indentation.
+                var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                var indentedText = GetIndentedText(indentation, documentOptions, formattedText);
+
+                // The caret should be placed at the end of the indented line.
+                var caretLocation = indentedText.Lines[desiredCaretLine.LineNumber].End;
+                var overallChanges = indentedText.GetTextChanges(originalDocumentText).ToImmutableArray();
+                return new BraceCompletionResult(overallChanges, caretLocation);
+            }
+
+            static TextLine GetLineBetweenCurlys(int closingPosition, SourceText text)
+            {
+                var closingBraceLineNumber = text.Lines.GetLineFromPosition(closingPosition - 1).LineNumber;
+                return text.Lines[closingBraceLineNumber - 1];
+            }
+
+            static SourceText GetIndentedText(IndentationResult indentation, DocumentOptionSet documentOptions, SourceText originalText)
+            {
+                var indentText = indentation.Offset.CreateIndentationString(documentOptions.GetOption(FormattingOptions.UseTabs), documentOptions.GetOption(FormattingOptions.TabSize));
+                var indentTextChange = new TextChange(new TextSpan(indentation.BasePosition, 0), indentText);
+                return originalText.WithChanges(indentTextChange);
+            }
+        }
+
+        public override async Task<bool> IsValidForBraceCompletionAsync(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
+        {
+            // Only potentially valid for curly brace completion if not in an interpolation brace completion context.
+            if (OpeningBrace == brace && await InterpolationBraceCompletionService.IsPositionInInterpolationContextAsync(document, openingPosition, cancellationToken).ConfigureAwait(false))
+            {
+                return false;
+            }
+
+            return await base.IsValidForBraceCompletionAsync(brace, openingPosition, document, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected override bool IsValidOpeningBraceToken(SyntaxToken token) => token.IsKind(SyntaxKind.OpenBraceToken) && !token.Parent.IsKind(SyntaxKind.Interpolation);
+
+        protected override bool IsValidClosingBraceToken(SyntaxToken token) => token.IsKind(SyntaxKind.CloseBraceToken);
+
+        private bool ContainsOnlyWhitespace(SourceText text, int openingPosition, int closingPosition)
+        {
+            var start = openingPosition;
+            start = text[start] == OpeningBrace ? start + 1 : start;
+
+            var end = closingPosition - 1;
+            end = text[end] == ClosingBrace ? end - 1 : end;
```

Comment:
this is so weird, esp. as you might have multiple of a closebrace next to each other.  i honestly don't get this code.

Reply:
Yep I didn't get this one either.  It's a straight copy but I'll see if I can clean it up